### PR TITLE
Heatmaps

### PR DESCRIPTION
**Pull request by ciremusyoka**
addressing issue #104

### CODE REVIEW

File: src/map/sortLayers.js [hunk 1]
```diff
@@ -12,121 +12,79 @@ export default function sortLayers(map, layers, nextLayerId) {
...
+        map.moveLayer(layers[key].id);
+      }
+    });
+    if (
+      layers.find(d => d.id === nextLayerId) &&
+      map.getLayer(nextLayerId)
+    ) {
+      map.moveLayer(nextLayerId);
+    }
+  }
```
Comment (cKellyDesign) [inline src/map/sortLayers.js]: Nice work making this reusable! 🎉

Comment (cKellyDesign) [reply src/map/sortLayers.js]: Please do add doc strings for this helper too

File: src/map/sortLayers.js [hunk 2]
```diff
@@ -12,106 +12,79 @@ export default function sortLayers(map, layers, nextLayerId) {
...
+    }
+
+    const heatMap = Object.keys(layers)
+      .map(d => layers[d])
+      .filter(d => d.type === 'heatmap');
+    if (heatMap.length) {
+      heatMap.sort((a,b) => b.priority - a.priority);
+      const checkPriority = Object.keys(heatMap).find(key => heatMap[key].priority)
+      if (checkPriority) {
+        Object.keys(heatMap).forEach((key) => {
```
Comment (cKellyDesign) [inline src/map/sortLayers.js]: Interesting, is `priority` a part of the layer specification?

File: src/map/sortLayers.js [hunk 3]
```diff
@@ -12,106 +12,79 @@ export default function sortLayers(map, layers, nextLayerId) {
...
+        map.moveLayer(filteredLayers[key].id);
+      }
+    });
+    if (
+      filteredLayers.find(d => d.id === nextLayerId) &&
+      map.getLayer(nextLayerId)
+    ) {
+      map.moveLayer(nextLayerId);
+    }
+  }
```
Comment (cKellyDesign) [inline src/map/sortLayers.js]: Nice work making this reusable! 🎉

Comment (cKellyDesign) [reply src/map/sortLayers.js]: Please do add doc strings for this helper too

**Comment by cKellyDesign** [comment]
@ciremusyoka it looks like there's an issue in the Legends component when rendering a heatmap layer after a fill layer (`stopsBreaks ` is undefined).

Screen Shot 2020-02-07 at 9 18 02 AM

File: src/map/sortLayers.js (hunk 2 above)
Comment (ciremusyoka) [reply src/map/sortLayers.js]: @cKellyDesign ~ When we have many `heatmap` layers and a user may want to view them at once it is impossible to know in which order the user will turn them on so as to adjust the colors and opacity of the layers. Hence in such cases `priority` is used to always arrange the layers in a specific order.

Comment (cKellyDesign) [reply src/map/sortLayers.js]: It feels like the order of the layers would normally be derived from the order the user turned on the individual layers. In the case of a grouped layer they should be rendered in the order they're set in the array.

Was this a specific request from the client?

Comment (ciremusyoka) [reply src/map/sortLayers.js]: Not really it is something we thought working on UNOPs but now they have gone back to using grouped layers so let me do away with it.

File: src/map/sortLayers.js [hunk 4]
```diff
@@ -10,108 +10,86 @@ export default function sortLayers(map, layers, nextLayerId) {
         6. cirle (other circle layers that aren't primary layers)
         7  line  primary
         8. line  (other line layers that aren't primary layers)
-        9. fill layers
+        9. heat map layers
+        19. fill layers
```
Comment (cKellyDesign) [inline src/map/sortLayers.js]: Should fill layers be `10` instead of `19`?

File: docs/layers/heatmap.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
+# HeatMaps
+
+This directory contains documentation on adding heatmap layers.\
+For full documentation follow this [link](https://docs.mapbox.com/help/tutorials/make-a-heatmap-with-mapbox-gl-js/).
```
Comment (cKellyDesign) [inline docs/layers/heatmap.md]: Nice! This is a great basic intro :)

File: docs/layers/heatmap.md [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    "label": "Sample Layer",
+    "type": "heatmap",
+    "source": {...},
+    "priority": <number>,
+    "paint": {
+        "heatmap-weight": [...]
+        "heatmap-radius": [...],
+        "heatmap-opacity": [...],
+        "heatmap-color": [...]
+    }
```
Comment (cKellyDesign) [inline docs/layers/heatmap.md]: Could we also include what are the default values for these? I think we can probably just copy and paste directly from the Mabbox docs.

File: docs/layers/heatmap.md [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+### paint properties
+
+For more details about the meaning and functions of each paint propery go to [Heatmap paint properties](https://docs.mapbox.com/help/tutorials/make-a-heatmap-with-mapbox-gl-js/#heatmap-paint-properties) on Mapbox.
+
+
+### priority property
+
+This is an optional property (a number) that can be used to give priorities on how heatmap layers are displayed on a map.
+When two or more layers are turned on at the same time, the layer with priority 1 will always be on top of the other layers. 
+
```
Comment (cKellyDesign) [inline docs/layers/heatmap.md]: Don't forget to remove it here too :)

File: docs/layers/heatmap.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+            0.8, 'rgb(239,138,98)',
+            1,  'rgb(178,24,43)'
+        ]
+    }
+}
+```
+
+### paint properties
+
+Note: The property `magnititude` in `heatmap-weight` can be any property from the dataset.
```
Comment (cKellyDesign) [inline docs/layers/heatmap.md]: `magnititude` => `magnitude`

File: docs/layers/heatmap.md [hunk 9]
```diff
@@ -0,0 +1,62 @@
+# HeatMaps
+
+This directory contains documentation on adding heatmap layers.\
+For full documentation follow this [link](https://docs.mapbox.com/help/tutorials/make-a-heatmap-with-mapbox-gl-js/).
+
+## Sample heatmap layer
+
```
Comment (cKellyDesign) [inline docs/layers/heatmap.md]: Maybe add a single line here saying that the values below are the default values? Not sure if that's true for `heatmap-weight`?

**Pull request merged**